crates/voicevox_core_c_api/src/object.rs

```diff
@@ -0,0 +1,145 @@
+use std::{
+    any,
+    collections::HashMap,
+    fmt::{Debug, Display},
+    mem,
+    ops::{Deref, DerefMut},
+    ptr::NonNull,
+    sync::Arc,
+};
+
+use easy_ext::ext;
+use tracing::warn;
+
+/// プロセスの終わりまでデストラクトされない、ユーザーにオブジェクトとして貸し出す1-bit長の構造体。
+///
+/// インスタンスは次のような形。
+///
+/// ```
+/// pub struct VoicevoxSynthesizer {
+///    _padding: MaybeUninit<[u8; 1]>,
+/// }
+/// ```
+///
+/// `RustApiObject`そのものではなくこのトレイトのインスタンスをユーザーに渡すようにすることで、次のことを実現する。
+///
+/// 1. "delete"時に対象オブジェクトに対するアクセスがあった場合、アクセスが終わるまで待つ
+/// 2. 次のユーザー操作に対するセーフティネットを張り、パニックするようにする
+///     1. "delete"後に他の通常のメソッド関数の利用を試みる
+///     2. "delete"後に"delete"を試みる
+///     3. そもそもオブジェクトとして変なダングリングポインタが渡される
+pub(crate) trait CApiObject: Default + Debug + 'static {
+    type RustApiObject: 'static;
+
+    // 書き込み操作としては`push`のみ
+    fn heads() -> &'static std::sync::Mutex<Vec<Self>>;
+
+    #[expect(
+        clippy::type_complexity,
+        reason = "型を分離するとかえって可読性を失う。その代わりコメントを入れている"
+    )]
+    fn bodies() -> &'static std::sync::Mutex<
+        HashMap<
+            usize, // `heads`の要素へのポインタのアドレス
+            Arc<
+                parking_lot::RwLock<
+                    Option<Self::RustApiObject>, // `RwLock`をdropする直前まで`Some`
+                >,
+            >,
+        >,
+    >;
+
+    fn new(body: Self::RustApiObject) -> NonNull<Self> {
+        assert!(mem::size_of::<Self>() > 0);
+
+        let this = {
+            let mut heads = Self::lock_heads();
+            heads.push(Default::default());
```

Comment:
I noticed that the `validate` function below checks if the pointer points inside `heads`, but this line can cause `heads` to reallocate, which means that creating a new object might invalidate existing pointers.

Reply:
Oh no thank you! You are obviously right.

I'll use [boxcar](https://docs.rs/crate/boxcar).

Reply:
(probably) fixed in #862.